server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -572,8 +573,13 @@ private void updateStaleCountOnCacheInsert(CleanupKey cleanupKey) {
 
         // pkg-private for testing
         void addToCleanupKeyToCountMap(ShardId shardId, String readerCacheKeyId) {
-            cleanupKeyToCountMap.computeIfAbsent(shardId, k -> ConcurrentCollections.newConcurrentMap())
-                .merge(readerCacheKeyId, 1, Integer::sum);
+            cleanupKeyToCountMap.compute(shardId, (currentShardId, readerCacheKeyMap) -> {
+                if (readerCacheKeyMap == null) {
+                    readerCacheKeyMap = new ConcurrentHashMap<>();
+                }
+                readerCacheKeyMap.compute(readerCacheKeyId, (currentReaderCacheKeyId, count) -> (count == null) ? 1 : count + 1);
+                return readerCacheKeyMap;
```

Comment:
Something of a style nitpick, but I try to avoid reassigning parameters in all cases:

```suggestion
                final ConcurrentHashMap<String, Integer> newValue;
                if (readerCacheKeyMap == null) {
                    newValue = new ConcurrentHashMap<>();
                } else {
                    newValue = readerCacheKeyMap;
                }
                newValue.compute(readerCacheKeyId, (currentReaderCacheKeyId, count) -> (count == null) ? 1 : count + 1);
                return newValue;
```

Reply:
Actually, I might say the previous implementation using `computeIfAbsent` and `merge` is more idiomatic (certainly more concise). I think it behaves differently too: this version nests one compute inside the other. In the case of ConcurrentHashMap these calls acquire locks, so I think it makes it _possible_ for this code to deadlock. I think you'd need some other code path that acquires locks in the reverse order for a deadlock to actually happen, but I think we might want to avoid these nested locks in any case.

Reply:
That's a fantastic point ! it makes sense to not change this part of the code then. I will revert these changes.

Reply:
By the way, the theoretical problem of deadlocks by using nested locks is an example of why I'd recommend avoiding this concurrentmap-within-a-concurrentmap pattern if at all possible. There are other code paths here where I think a compute is happening within a compute, but I think the order of lock acquisition will always be the same so it seems like it should not deadlock (unless future refactorings change things).